File: Leetcode/12_best_item_to_buy_and_sell_stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let profit = 0;
  let minValue = Number.MAX_VALUE;

  for (const price of prices) {
    minValue = Math.min(price, minValue);
    profit = Math.max(profit, price - minValue);
  }

  return profit;
};

console.log(maxProfit([7, 1, 5, 3, 6, 4]));

File: Leetcode/12_best_item_to_buy_and_sell_stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
from typing import List
from sys import maxsize


def maxProfit(prices: List[int]) -> int:
    profit = 0
    min_price = maxsize

    # 최소값과 최대값을 계속 갱신
    for price in prices:
        min_price = min(price, min_price)
        profit = max(profit, price - min_price)

    return profit


print(maxProfit([7, 1, 5, 3, 6, 4]))
